File: german-telework/service/messagecrypt/src/transport_crypt.cc
#include "transport_crypt.hh"

uint64_t transport_crypt_keystream(uint64_t seed) {
	uint64_t result = seed;
	result ^= result << 13;
	result ^= result >> 7;
	result ^= result << 17;
	return result;
}

uint8_t transport_crypt_get_key_byte(transport_crypt_state_t* state) {
	if (state->count == 16) {
		state->count = 0;
		state->seed = transport_crypt_keystream(state->seed);
	}
	return (state->seed >> (state->count)++);
}

void transport_crypt_state(transport_crypt_state_t* state, uint8_t* buf, size_t n) {
	for (size_t i = 0; i < n; i++) {
		uint8_t key_byte = transport_crypt_get_key_byte(state);
		buf[i] ^= key_byte;
	}
}

ssize_t recv_tc_state(transport_crypt_state_t* state, int sockfd, void* buf, size_t len, int flags) {
	ssize_t bytes_recvd = recv(sockfd, buf, len, flags);
	if (bytes_recvd > 0) {
		transport_crypt_state(state, (uint8_t*)buf, bytes_recvd);
	}
	return bytes_recvd;
}

ssize_t recv_tc(int sockfd, void* buf, size_t len, int flags) {
	transport_crypt_state_t state;
	return recv_tc_state(&state, sockfd, buf, len, flags);
}

ssize_t send_tc_state(transport_crypt_state_t* state, int sockfd, void const* buf, size_t len, int flags) {
	transport_crypt_state(state, (uint8_t*)buf, len);
	return send(sockfd, buf, len, flags);
}

ssize_t send_tc(int sockfd, void const* buf, size_t len, int flags) {
	transport_crypt_state_t state;
	return send_tc_state(&state, sockfd, buf, len, flags);
}

File: german-telework/service/gateway/transport_crypt.py
TRANSPORT_CRYPT_INITIAL_SEED = 0xEA0A8EEA0A8EEA0A
U64MASK = 0xffffffffffffffff

def _transport_crypt_keystream(seed: int) -> int:
	result = seed & U64MASK
	result ^= result << 13 & U64MASK
	result ^= result >> 7 & U64MASK
	result ^= result << 17 & U64MASK
	return result & U64MASK

class TransportCryptState:
	def __init__(self) -> None:
		self.seed: int = TRANSPORT_CRYPT_INITIAL_SEED
		self.count: int = 0

	def _transport_crypt_get_key_byte(self) -> int:
		if self.count == 16:
			self.count = 0
			self.seed = _transport_crypt_keystream(self.seed)
		byte = (self.seed >> self.count) & 0xff
		self.count += 1
		return byte

	def transport_crypt(self, buf: bytearray) -> bytearray:
		for i in range(len(buf)):
			buf[i] ^= self._transport_crypt_get_key_byte()
		return buf

def transport_crypt(buf: bytearray) -> bytearray:
	state = TransportCryptState()
	buf = state.transport_crypt(buf)
	return buf